kore/src/Kore/Rewrite/Axiom/Matcher.hs

```diff
@@ -81,696 +68,401 @@ import Kore.Internal.Predicate (
 import Kore.Internal.Predicate qualified as Predicate
 import Kore.Internal.SideCondition
 import Kore.Internal.SideCondition qualified as SideCondition
-import Kore.Internal.Symbol qualified as Symbol
 import Kore.Internal.TermLike
 import Kore.Rewrite.Axiom.MatcherData
 import Kore.Rewrite.RewritingVariable (
     RewritingVariableName,
  )
 import Kore.Simplify.InjSimplifier as InjSimplifier
 import Kore.Simplify.Overloading (
-    matchOverloading,
+    OverloadingResolution (..),
+    flipResult,
+    unifyOverloadingCommonOverload,
+    unifyOverloadingVsOverloaded,
+ )
+import Kore.Simplify.Overloading qualified as Overloading (
+    MatchResult (..),
  )
 import Kore.Simplify.Simplify (
     Simplifier,
  )
 import Kore.Simplify.Simplify qualified as Simplifier
-import Kore.Substitute qualified as Substitute
-import Kore.Variables.Binding
-import Kore.Variables.Fresh qualified as Variables
+import Kore.Substitute
+import Kore.Unparser (
+    unparseToText,
+ )
 import Pair
 import Prelude.Kore
 
 -- * Matching
 
-newtype Constraint variable = Constraint
-    { getConstraint :: Pair (TermLike variable)
-    }
-    deriving stock (Eq)
+{- | Match a TermLike against a pattern as represented by another TermLike.
```

Comment:
I appreciate this documentation, but I think it would be more helpful to add comments to the other functions in this file as well. Just a few lines describing what they do, and how they fit into this algorithm.